server/src/test/java/org/elasticsearch/indices/recovery/DanglingIndicesIT.java

```diff
@@ -126,29 +146,291 @@ public Settings onNodeStopped(String nodeName) throws Exception {
+        // amount of time
+        assertFalse(
+            "Did not expect dangling index " + INDEX_NAME + " to be recovered",
+            waitUntil(() -> indexExists(INDEX_NAME), 1, TimeUnit.SECONDS)
+        );
+    }
+
+    /**
+     * Check that when dangling indices are discovered, then they can be listed.
+     */
+    public void testDanglingIndicesCanBeListed() throws Exception {
+        internalCluster().startNodes(3, buildSettings(0, true, false));
+
+        final String stoppedNodeName = createDanglingIndices(INDEX_NAME);
+
+        final ListDanglingIndicesResponse response = client().admin()
+            .cluster()
+            .listDanglingIndices(new ListDanglingIndicesRequest())
+            .actionGet();
+        assertThat(response.status(), equalTo(RestStatus.OK));
+
+        final List<NodeListDanglingIndicesResponse> nodeResponses = response.getNodes();
+        assertThat("Didn't get responses from all nodes", nodeResponses, hasSize(3));
+
+        for (NodeListDanglingIndicesResponse nodeResponse : nodeResponses) {
+            if (nodeResponse.getNode().getName().equals(stoppedNodeName)) {
+                assertThat("Expected node that was stopped to have one dangling index", nodeResponse.getDanglingIndices(), hasSize(1));
+
+                final DanglingIndexInfo danglingIndexInfo = nodeResponse.getDanglingIndices().get(0);
+                assertThat(danglingIndexInfo.getIndexName(), equalTo(INDEX_NAME));
+            } else {
+                assertThat("Expected node that was never stopped to have no dangling indices", nodeResponse.getDanglingIndices(), empty());
+            }
+        }
+    }
+
+    /**
+     * Check that dangling indices can be imported.
+     */
+    public void testDanglingIndicesCanBeImported() throws Exception {
+        internalCluster().startNodes(3, buildSettings(0, true, false));
+
+        final String stoppedNodeName = createDanglingIndices(INDEX_NAME);
+
+        final String danglingIndexUUID = findDanglingIndexForNode(stoppedNodeName, INDEX_NAME);
+
+        final ImportDanglingIndexRequest request = new ImportDanglingIndexRequest(danglingIndexUUID, true);
+
+        client().admin().cluster().importDanglingIndex(request).actionGet();
+
+        assertTrue("Expected dangling index " + INDEX_NAME + " to be recovered", indexExists(INDEX_NAME));
+    }
+
+    /**
+     * Check that the when sending an import-dangling-indices request, the specified UUIDs are validated as
+     * being dangling.
+     */
+    public void testDanglingIndicesMustExistToBeImported() {
+        internalCluster().startNodes(1, buildSettings(0, true, false));
+
+        final ImportDanglingIndexRequest request = new ImportDanglingIndexRequest("NonExistentUUID", true);
+
+        boolean noExceptionThrown = false;
+        try {
+            client().admin().cluster().importDanglingIndex(request).actionGet();
+            noExceptionThrown = true;
```

Comment:
Could we `expectThrows(IllegalArgumentException.class, ...)` here instead of this?